datafusion/src/logical_plan/builder.rs

```diff
@@ -1395,4 +1508,58 @@ mod tests {
         assert!(stringified_plan.should_display(true));
         assert!(!stringified_plan.should_display(false));
     }
+
+    #[test]
+    fn invalid_plan_with_not_non_boolean_expression() -> Result<()> {
+        let mut plan = LogicalPlanBuilder::scan_empty(
+            Some("employee_csv"),
+            &employee_schema(),
+            Some(vec![0, 1, 2, 3, 4]),
+        )?;
+        plan = plan.project(vec![Expr::Not(Box::new(Expr::Column(Column {
+            relation: None,
+            name: "id".to_string(),
+        })))])?;
```

Comment:
I think the "builder" API for expressions makes for much easier reading:

```suggestion
        plan = plan.project(vec![col("id").not()])?;
```